plugins/chat/app/services/lookup_thread.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Chat
+  module Service
+    # Finds a thread within a channel. The thread_id and channel_id must
+    # match. For now we do not want to allow fetching threads if the
+    # enable_experimental_chat_threaded_discussions hidden site setting
+    # is not turned on, and the channel must specifically have threading
+    # enabled.
+    #
+    # @example
+    #  Chat::Service::LookupThread.call(thread_id: 88, channel_id: 2, guardian: guardian)
+    #
+    class LookupThread
+      include Base
+
+      # @!method call(thread_id:, channel_id:, guardian:)
+      #   @param [Integer] thread_id
+      #   @param [Integer] channel_id
+      #   @param [Guardian] guardian
+      #   @return [Chat::Service::Base::Context]
+
+      policy :threaded_discussions_enabled
+      contract
+      model :thread, :fetch_thread
+      policy :invalid_access
+      policy :threading_enabled_for_channel
+
+      # @!visibility private
+      class Contract
+        attribute :thread_id, presence: true
+        attribute :channel_id, presence: true
+      end
+
+      private
+
+      def threaded_discussions_enabled
+        SiteSetting.enable_experimental_chat_threaded_discussions
+      end
+
+      def fetch_thread(thread_id:, channel_id:, **)
+        ChatThread
+          .includes(:channel)
+          .includes(original_message_user: :user_status)
+          .includes(original_message: :chat_webhook_event)
+          .find_by!(id: thread_id, channel_id: channel_id)
```

Comment:
We should use `find_by` instead of `find_by!` as the latter will raise an exception that won’t be caught by the service. While this is working in the controller calling this service because we have a `rescue_from ActiveRecord::RecordNotFound` defined somewhere, if we were to call the service in another context, the exception will have to be handled manually. (instead of just being able to inspect the result object)
We’ll have to add this line of code in the controller though:
```ruby
      on_model_not_found(:thread) { raise Discourse::NotFound }
```
This way, everything is “self-contained” inside the service.

Reply:
I will change this one, but I do wonder why we don't just handle `ActiveRecord::NotFound` inside the `ModelStep` class, same as how we rescue `ArgumentError`? Then whoever is writing the service doesn't need to worry about it.

https://github.com/discourse/discourse/blob/1eaebd686439e0a2b32bbbd5f07fc7e32f3efca6/plugins/chat/app/services/base.rb#L216-L225

Reply:
Also why raise `ArgumentError` here instead of just `Discourse::NotFound`?

Reply:
The idea is that a service doesn't rely on exceptions for control flow and if an exception were to be raised, then it will interrupt the sequence of steps leaving the service in an unknown/unstable state. And this is good for catching bugs  :grin: 
I remember @jjaffeux having the same thought as you about AR exceptions but dropped it in the end. He will probably explain this better than me.

Here, the `ArgumentError` handling is something we did to prevent the service from breaking when a parameter was missing. As calling the `model` step first without checking for parameters validity before is something that seemed totally fine to do (as in the end we’re just passing one argument to `#find_by` typically), it wouldn’t have been nice to have the whole service break just for a missing argument. Maybe that’s something we should revisit though. We could decide to remove this `ArgumentError` logic altogether and to rely on a contract instead. We could also simply decide that if an expected parameter isn’t present, then we should treat that as a bug and not as something expected. (typically when the `id` comes from a Rails route and not from a query arg, we can treat it as a dependency instead of an input parameter)

Reply:
Yep thanks for the explanation. I do remember the control flow discussions, I mostly meant why `raise ArgumentError` here instead of `raise Discourse::NotFound`. All good I see the intent now.